common/phal_F4/rcc/rcc.c

```diff
@@ -0,0 +1,366 @@
+{
+    // Valid range for PLL output is 64Mhx <= PLL_CLK <= 344Mhz
+    vco_output_rate_target_hz = vco_output_rate_target_hz > (uint32_t) 344e6 ? (uint32_t) 344e6 : vco_output_rate_target_hz;
+    vco_output_rate_target_hz = vco_output_rate_target_hz < (uint32_t) 64e6 ? (uint32_t) 64e6 : vco_output_rate_target_hz;
+
+    RCC->CR &= ~RCC_CR_PLLON;
+    while((RCC->CR & RCC_CR_PLLRDY))
+        ; // Wait for PLL to turn off
+
+    RCC->PLLCFGR &= ~(RCC_PLLCFGR_PLLSRC_Msk | RCC_PLLCFGR_PLLN_Msk | RCC_PLLCFGR_PLLM_Msk);
+
+    /* Determine input clock rate from the selected source */
+    uint32_t pll_input_f_hz;
+    switch (pll_source)
+    {
+        case PLL_SRC_HSI16:
+            RCC->PLLCFGR |= RCC_PLLCFGR_PLLSRC_HSI;
+            RCC->CR |= RCC_CR_HSION;
+            while(!(RCC->CR & RCC_CR_HSIRDY))
+                ; // Wait for HSI to turn on
+            pll_input_f_hz = HSI_CLOCK_RATE_HZ;
+            break;
+
+        case PLL_SRC_MSI:
+            RCC->PLLCFGR |= RCC_PLLCFGR_PLLSRC_HSI;
+            pll_input_f_hz = MSIClockRateHz;
+            break;
+
+        case PLL_SRC_HSE:
+            if (HSE_CLOCK_RATE_HZ == HSE_CLOCK_RATE_HZ_INVALID)
+            {
+                return false; // Make sure you set the HSE_CLOCK_RATE_HZ to what is on the PCB!
+            }
+            RCC->PLLCFGR |= RCC_PLLCFGR_PLLSRC_HSE;
+            pll_input_f_hz = HSE_CLOCK_RATE_HZ;
+            break;
+    }
+
+    /* Search for a possible PLL configuration */
+    uint8_t pll_input_divisor = 1;
+    uint8_t pll_output_multiplier = 8;
+    bool valid_rate = false;
+    for (; pll_input_divisor <= 8; pll_input_divisor ++)
+    {
+        // VCO input clock is only allowed between 4MHz and 16MHz
+        uint32_t pll_vco_in_rate = pll_input_f_hz / pll_input_divisor;
+        if(pll_vco_in_rate < 4000000 || pll_vco_in_rate > 16000000)
+        {
+            continue;
+        }
+
+        for (; pll_output_multiplier <= 86; pll_output_multiplier ++)
+        {
+            if ((pll_input_f_hz / pll_input_divisor) * pll_output_multiplier == vco_output_rate_target_hz)
+            {
+                valid_rate = true;
+                break;
+            }
+        }
+        if (valid_rate)
+            break;
+    }
+
+    if (!valid_rate)
+    {
+        return false; // Unable to find a valid clock rate!
+    }
+
+    RCC->PLLCFGR |= ((pll_input_divisor - 1) << RCC_PLLCFGR_PLLM_Pos) & RCC_PLLCFGR_PLLM_Msk;
+    RCC->PLLCFGR |= ((pll_output_multiplier) << RCC_PLLCFGR_PLLN_Pos) & RCC_PLLCFGR_PLLN_Msk;
+
+    // Update global variable used to reference the PLL
+    PLLClockRateHz = (pll_input_f_hz / pll_input_divisor) * pll_output_multiplier;
+
+    RCC->CR |= RCC_CR_PLLON;
+    while(!(RCC->CR & RCC_CR_PLLRDY))
+        ; // Wait for PLL to turn on
+
+    SystemCoreClockUpdate();
+    return true;
+}
+
+// bool PHAL_configurePLLSystemClock(uint32_t system_clock_target_hz)
```

Comment:
what's with the commented out function

Reply:
The RCC HAL is currently being worked on, but this branch is simply to make the build system work with the F4. The RCC HAL is good enough to blink the LED to show the efficacy of the build system, but I will actually make it better. The PLL clock control from L4 to F4 is very different, so I am working on that currently